algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,316 @@
+/- 
+Copyright (c) 2018 Louis Carlin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Louis Carlin
+
+Euclidean domains and Euclidean algorithm (extended to come)
+A lot is based on pre-existing code in mathlib for natural number gcds
+-/
+import tactic.ring
+
+universe u
+
+class euclidean_domain (α : Type u) extends integral_domain α :=
+(quotient : α → α → α)
+(remainder : α → α → α)
+(witness : ∀ a b, (quotient a b) * b + (remainder a b) = a) -- This could be changed to the same order as int.mod_add_div. We normally write qb+r rather than r + qb though.
+(valuation : α → ℕ)
+(val_mod : ∀ a b, b = 0 ∨  valuation (remainder a b) <  valuation b)
```

Comment:
there are stil strange spaces in `val_mod`.

Reply:
Change the disjunction to a implication `b /= 0 -> ...`

Also it should be `valuation_remainder_lt`.